### DISCUSSION TITLE
Set up integration test setup only once

### DISCUSSION THREAD

**Discussion by kevcodez** [vitest-dev/vitest#1710] [Q&A]
Hi there,

I was wondering if it's possible to do a global setup for integration tests ONCE only. I've struggled to do this with Jest, as it seems super unreliable to do so.

What I would like to achieve:

Separate unit and integration tests (already done so by giving it a different file name pattern), run the unit tests just as they come, but for the integration tests, run the setup (starting in-memory Mongo, Redis, ...) only once and share the setup and eventually have a global teardown method taking care of the teardown afterwards.

In jest, this is supposed to be possible with testEnvironments, but they never worked for me, so I was wondering whether Vitest supports a similar approach that would enable a one-time setup/teardown (not per file, but for the entire test suite).

**Comment by danielhra** [comment]
Did you find a way to achieve this?

**Reply by kevcodez** [reply]
Still stuck on Jest and suffering in silence

**Comment by AriPerkkio** [comment]
> I was wondering whether Vitest supports a similar approach that would enable a one-time setup/teardown (not per file, but for the entire test suite).

Sounds exactly like a use-case for [globalSetup](https://vitest.dev/config/#globalsetup). Does that work for you?

**Reply by peacechen** [reply]
`globalSetup` runs for each integration test file. This is a big problem if the tests are run in parallel. There needs to be a way to run a single initialization instance for all the tests. 

`isolate: false` doesn't help unfortunately. The only way I've gotten it to work is to set `fileParallelism: false` which slows down the tests a lot.

**Reply by AriPerkkio** [reply]
Without using `fileParallelism: false`, you can do this with custom reporter:

```ts
import { defineConfig } from 'vitest/config'

export default defineConfig({
  test: {
    reporters: {
      onInit() {
        // run setup here
      },
    },
  },
})
```

**Reply by sheremet-va** [reply]
`globalSetup` doesn't run for each integration test file. `globalSetup` runs once for a project it is defined in. `setupFiles` run for every test file.

**Reply by peacechen** [reply]
> `globalSetup` doesn't run for each integration test file. `globalSetup` runs once for a project it is defined in. `setupFiles` run for every test file.

That may be the intent, but it runs for every test file. I have test suites in separate `.test.ts` files and this executes repeatedly, once for each test file:
```
export default <Environment>{
  name: "prisma",
  transformMode: "ssr",
  async setup() {
    const env = { ... };

    await execSync(
      "npx prisma db push --force-reset",
      { env },
    );

    return {
      async teardown() {
        // called after all tests with this env have been run
        await prisma.$disconnect();
      },
    };
  },
};
```

When run in parallel, every test tries to clear the db, causing the other tests to fail.

**Reply by sheremet-va** [reply]
This is not a global setup, this is a [custom environment](https://vitest.dev/config/#environment). Environments are created for every file unless isolation is disabled.

**Reply by CleytonGoncalves** [reply]
`globalSetup` indeed runs once per project. Is there a way to make it truly global? i.e., once per vitest execution, not just per project?

**Reply by AriPerkkio** [reply]
@CleytonGoncalves a custom reporter should be able to do that.
- https://vitest.dev/api/advanced/reporters.html

**Reply by peacechen** [reply]
> @CleytonGoncalves a custom reporter should be able to do that.
> 
> * https://vitest.dev/api/advanced/reporters.html

Thanks for the work-around. Reporters look like they could be leveraged for lots of different uses. Semantically it isn't intuitive though. For someone looking in the documentation for a way to hook into a single global setup, I would never have thought to look at reporters.

**Comment by yudin-s** [comment]
Yes, `globalSetup` is the thing you want for this. It runs once before the test workers start, and the function it returns is used as teardown after the suite finishes.

For something like Mongo, I would keep it on a separate integration-test config or Vitest project so unit tests do not pay the startup cost:

```ts
// vitest.config.ts
import { defineConfig } from "vitest/config"

export default defineConfig({
  test: {
    include: ["**/*.integration.test.ts"],
    globalSetup: ["./test/global-setup.ts"],
  },
})
```

```ts
// test/global-setup.ts
import type { TestProject } from "vitest/node"

export default async function setup(project: TestProject) {
  const mongo = await startMongoSomehow()

  project.provide("mongoUri", mongo.uri)

  return async () => {
    await mongo.stop()
  }
}

declare module "vitest" {
  export interface ProvidedContext {
    mongoUri: string
  }
}
```

And in a test:

```ts
import { inject } from "vitest"

const mongoUri = inject("mongoUri")
```

I would avoid relying on plain globals here because setup runs outside the test file scope. `project.provide()` / `inject()` makes the handoff explicit.
